src/main/java/io/cryostat/rules/RuleMatcher.java

```diff
@@ -0,0 +1,139 @@
+            evt.begin();
+            Object result =
+                    this.scriptEngine.eval(rule.getMatchExpression(), createBindings(serviceRef));
+            if (result instanceof Boolean) {
+                return (Boolean) result;
+            } else {
+                throw new ScriptException(
+                        String.format(
+                                "Rule %s non-boolean match expression evaluation result: %s",
+                                rule.getName(), result));
+            }
+        } finally {
+            evt.end();
+            if (evt.shouldCommit()) {
+                evt.commit();
+            }
+        }
+    }
+
+    Bindings createBindings(ServiceRef serviceRef) {
+        BindingsCreationEvent evt = new BindingsCreationEvent();
+        try {
+            evt.begin();
+            Bindings bindings = this.scriptEngine.createBindings();
+            Map<String, String> cryostatAnnotations =
+                    new HashMap<>(serviceRef.getCryostatAnnotations().size());
+            for (Map.Entry<ServiceRef.AnnotationKey, String> entry :
+                    serviceRef.getCryostatAnnotations().entrySet()) {
+                cryostatAnnotations.put(entry.getKey().name(), entry.getValue());
+            }
+            bindings.put(
+                    "target",
+                    Map.of(
+                            "connectUrl",
+                            serviceRef.getServiceUri(),
+                            "alias",
+                            serviceRef.getAlias().orElse(null),
+                            "labels",
+                            serviceRef.getLabels(),
+                            "annotations",
+                            Map.of(
+                                    "platform",
+                                    serviceRef.getPlatformAnnotations(),
+                                    "cryostat",
+                                    cryostatAnnotations)));
+            return bindings;
+        } finally {
+            evt.end();
+            if (evt.shouldCommit()) {
+                evt.commit();
+            }
+        }
+    }
+
+    @Name("io.cryostat.rules.RuleMatcher.RuleAppliesEvent")
+    @Label("Rule Expression Matching")
+    @Category("Cryostat")
+    @SuppressFBWarnings(
+            value = "URF_UNREAD_FIELD",
+            justification = "The event fields are recorded with JFR instead of accessed directly")
+    public static class RuleAppliesEvent extends Event {
+
+        String ruleName;
+
+        RuleAppliesEvent(String ruleName) {
+            this.ruleName = ruleName;
+        }
+    }
+
+    @Name("io.cryostat.rules.RuleMatcher.BindingsCreationEvent")
+    @Label("Rule Binding Creation")
+    @Category("Cryostat")
+    @SuppressFBWarnings(
+            value = "URF_UNREAD_FIELD",
+            justification = "The event fields are recorded with JFR instead of accessed directly")
+    public static class BindingsCreationEvent extends Event {}
```

Comment:
Are these events meant to be expanded in the future? Right now it seems like the BindingsCreationEvent always occurs at the same time as the RuleAppliesEvent, but doesn't provide any additional info.

Reply:
The `RuleAppliesEvent` starts strictly before and ends strictly after the `BindingsCreationEvent`, with a 1:1 relationship. I meant for these events to be used for profiling how long Nashorn evaluation of match expressions takes, and also to see how much of that time is actually evaluation by the rule engine vs. time spent in Cryostat preparing the bindings to pass to the engine. The time taken by Nashorn evaluation is roughly the duration of the `RuleAppliesEvent` minus the duration of the `BindingsCreationEvent`.

Reply:
Ah, I understand now. I hadn't considered it being a duration thing.